solarkraft/src/fetch.ts

```diff
@@ -10,64 +10,90 @@
+        lastHeight = cachedHeight
+    } else {
+        lastHeight = args.height
+    }
 
-    const DURATION = 30
-    const server = new Horizon.Server(args.rpc)
-    console.log(`Fetching the ledger for ${startHeight}`)
-    const response = await server.ledgers().ledger(startHeight).call()
-    //console.log(response)
+    console.log(`Fetching fresh transactions from: ${args.rpc}...`)
+    
+    console.log(`Fetching the ledger for ${lastHeight}`)
+    const response = await server.ledgers().ledger(lastHeight).call()
     const startCursor = (response as any).paging_token
+    // timeout id, if a timeout is set below
+    let timeout
 
-    // See:
-    // https://github.com/stellar/js-stellar-sdk/blob/master/test/integration/streaming_test.js
-    const finish = (err) => {
+    const done = (err) => {
+        // Closing handler:
+        // https://github.com/stellar/js-stellar-sdk/blob/master/test/integration/streaming_test.js
         clearTimeout(timeout)
         closeHandler()
         console.error(err)
     }
 
-    // TODO: work in progress
+    // the number of received events
+    let nEvents = 0
+
+    // initiate the streaming loop
     const closeHandler = server
         .operations()
         .cursor(startCursor)
         .stream({
             onmessage: async (msg: any) => {
-                if (msg.transaction_successful === true) {
-                    ;(
-                        await extractContractCall(
-                            msg,
-                            (id) => contractId === id
-                        )
-                    ).map((e) => {
-                        console.log(`call => ${JSONbigint.stringify(e)}`)
-                    })
+                if (msg.transaction_successful) {
+                    const callEntryMaybe =
+                        await extractContractCall(msg, (id) => contractId === id)
+                    if (callEntryMaybe.isJust()) {
+                        const entry = callEntryMaybe.value
+                        console.log(`+ save: ${entry.height}`)
+                        saveContractCallEntry(args.home, entry)
+                    }
+                } // else: shall we also store reverted transactions?
+
+                nEvents++
+                if (nEvents % HEIGHT_FETCHING_PERIOD === 0) {
+                    // Fetch the height of the current message.
+                    // Note that messages may come slightly out of order, so the heights are not precise.
+                    const tx = await msg.transaction()
+                    lastHeight = Math.max(lastHeight, tx.ledger_attr)
+                    console.log(`= at: ${lastHeight}`)
+                    // load and save the state, other fetchers may work concurrently
+                    fetcherState = loadFetcherState(args.home)
+                    fetcherState = { ...fetcherState, heights: fetcherState.heights.set(contractId, lastHeight) }
+                    saveFetcherState(args.home, fetcherState)
```

Comment:
Don't you get a filesystem race here with concurrent fetchers? 😄 

And actually, will the streaming be sequential?

Reply:
yeah, there is a possibility into missing filesystem updates for concurrently running fetchers. Also, for the callbacks in theory.